### DISCUSSION TITLE
Difference between -n=auto and -n=12 (I have 12 logical cores)

### DISCUSSION THREAD

**Discussion by samuelwhiskeyjohnson** [pytest-dev/pytest-xdist#1107] [General]
-n=auto uses as many CPU cores as possible. But why is there difference with manually specifying the number by yourself e.g.) -n=12 when you have 12 logical cores.

**Comment by RonnyPfannschmidt** [comment]
It's not clear what the ask is

**Reply by samuelwhiskeyjohnson** [reply]
Why is there difference in performance between -n=auto and -n=12 (12 being the maximum cores I have)?

**Reply by RonnyPfannschmidt** [reply]
What n is getting picked automatically?

**Reply by samuelwhiskeyjohnson** [reply]
I'm not sure what the confusion is? Isn't this the pytest plugin where you can run pytest in parallel?

In the book, Python Testing with Pytest,
`pytest --count=10 -n=auto test_parallel.py`
This is how you would run pytest on test_parallel.py 10 times. `-n=auto` makes the pytest run in parallel to run on as many CPU cores as possible.

But you can also manually specify the CPU cores by `-n=<what ever number>` such as `-n=6`

**Comment by RonnyPfannschmidt** [comment]
closing as the question left out details on whats happening and we cant answer without guessing everything

**Reply by samuelwhiskeyjohnson** [reply]
Can you at least give me the details you ask for? At least I gave the effort to answer your question. 
What part of my question didn't make sense? I genuinely feel I explained explicitly. I don't know how it can be more explained more clearly.

In the documentation it says
With `-n auto`, pytest-xdist will use as many processes as your computer has physical CPU cores.

So my question is why is there performance difference between `-n auto` and `-n 8` when i have 8 cores? Intuitively there shouldn't be a performance difference. e.g.) Why is `-n auto` take 5 seconds to complete the tests when `-n 8` takes 7 seconds to complete the tests

**Reply by webknjaz** [reply]
You have to check the logs and see how many cores it selects in the automatic mode. It might go for the logical core number, not physical, for example. In some environments, the detection may return incorrect numbers (I think, this was observed in some virtualisation modes on CIs). So on CIs, people often set a lower number, keeping auto in the shared confign for any other invocations.
What's the output of `cat /proc/cpuinfo` and `nproc`?

**Reply by webknjaz** [reply]
One more point: spawning subprocesses is expensive. So if you only have a few tests, oftentimes it's faster without xdist at all. You need to observe what's happening in your case.

**Reply by RonnyPfannschmidt** [reply]
@samuelwhiskeyjohnson if the same number of processes is used, there shouldn't be a actual difference - if there is, its a bug that needs a reproducer

**Reply by samuelwhiskeyjohnson** [reply]
> What's the output of `cat /proc/cpuinfo` and `nproc`?

I'm on windows so this command doesnt work. But I have Ryzen 3600 which has 6 Physical cores and 12 Logical cores. 
so `pytest -n auto` and `pytest -n 12` gives you the same time. 


I think I made a mistake. What I meant to say is there is a difference between `pytest -n auto` and `pytest -n 6`.
I did `6` because when I read the [documentation](https://github.com/pytest-dev/pytest-xdist/commit/f57c65839bccedd4f9fcad4c24d778d3fb3f8c2a#diff-6b91d225982ba93705c6cf4a26bf6e347a98b811cf39518569dd280628edc0dfR13) is says `With ``-n auto``, pytest-xdist will use as many processes as your computer has physical CPU cores.` 

But isn't this wrong? You specify the number of logical cores, not physical cores. When I run `pytest -n auto` it gives you 12 workers like there are 12 logical cores as Ryzen 3600 has. 
If the documentation is right then `pytest -n auto` should give you 6 workers as there are 6 physical cores for Ryzen 3600.

<img width="543" alt="image" src="https://github.com/user-attachments/assets/bd5a4226-7c3b-4a2a-869f-c6192c1759c0">

**Reply by webknjaz** [reply]
@RonnyPfannschmidt I suppose this might be a docs bug, right? Having looked at https://pytest-xdist.rtfd.io/en/latest/distribution.html#running-tests-across-multiple-cpus, it seems that in this scenario, `-n auto` would have to return 6 and `-n logical` would be 12. However, it's probably the opposite.

I tried the same detection techniques locally, and here's what I got:

```pycon
$ python
Python 3.11.4 (main, Jul 18 2023, 13:30:02) [GCC 12.2.1 20230428] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import psutil
>>> psutil.cpu_count()
12
>>> psutil.cpu_count(logical=True)
12
>>> psutil.cpu_count(logical=False)
6
>>> from os import sched_getaffinity
>>> sched_getaffinity(0)
{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}
>>> len(sched_getaffinity(0))
12
>>> from os import cpu_count
>>> cpu_count()
12
>>> from multiprocessing import cpu_count
>>> cpu_count()
12
```

This essentially means that `-n auto` returns whatever the kernel/OS sees, which is usually the number of the logical cores. Except for when `psutil` is installed — in this case, `config.option.numprocesses == "logical"` will be `False` and then, `-n auto` would return the number of physical cores (https://github.com/pytest-dev/pytest-xdist/blob/d8efeb9/src/xdist/plugin.py#L31).
The `-n logical` path seems to hit the right branch and return the number of logical cores always.

So I think that the docs might indeed be wrong. I wonder if it'd make sense to have an `-n physical` too.

**Reply by nicoddemus** [reply]
> So I think that the docs might indeed be wrong. 

I think you are right, thanks for the detailed look.
